design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,315 @@
+
+
+## Goals
+Design and develop an API to handle storage migration between PVCs.
+
+ ## Non  Goals
+*  This features is not directly aimed for live migrating the VM using RWO volumes. Hence, the API covers and refers only to the volume migration.
+*  This design doesn't cover migration between storage classes, but rather implements a basic APIs that can be used by overlying tooling
+*  This design doesn't include the creation of the destination PVCs, it expects the destination PVCs to exist.
+
+ ## Definition of Users
+This feature addresses for all KubeVirt users who want to move the PVCs assigned to a VM without workload interruptions.
+
+## User Stories
+
+  1. As a VM owner, the storage class of volumes used by a running VM is deprecated, and the volumes need to be moved to another storage class
+  2. As a VM owner, I want to use higher performance storage class that is now available in the cluster
+  3. As a VM owner, I see my non-resizable storage device filling up and I want to move my VM to a larger storage device without interruption in order to give it more storage
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec VolumeMigrationSpec `json:"spec"`
+	Status *VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items []VolumeMigration `json:"items"`
+}
+
+type ReclaimPolicySourcePvc string
+
+const (
+	DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+	RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+	DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+	ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty" valid:"required"`
+}
+
+type VolumeMigrationSpec struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type RejectedMigratedVolume struct {
+	SourcePvc string `json:"sourcePvc,omitempty" valid:"required"`
+	Reason    string `json:"reason,omitempty" valid:"required"`
+}
+
+type RejectedVolumeMigration struct {
+	RejectedVolumes []RejectedMigratedVolume `json:"rejectedVolumes,omitempty" valid:"required"`
+	AssociatedVolumes []string `json:"associatedVolumes,omitempty"`
+	Reason            string   `json:"reason,omitempty" valid:"required"`
```

Comment:
We already have `Reason` in each `RejectedVolumes` item. Could you give an example how this one is used?

Reply:
Yes, I left it on both places on purpose because we can have a reason only specific to a single disk (e.g shareable, hotplug...) or it can be related to the entire migration like the volumes belong to different VMIs.

Reply:
Here an example:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: vm1
spec:
  template:
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: datavolumedisk
      volumes:
      - dataVolume:
          name: src-pvc1
        name: datavolumedisk
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: vm2
spec:
  template:
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: datavolumedisk
      volumes:
      - dataVolume:
          name: src-pvc2
        name: datavolumedisk
---
apiVersion: storage.kubevirt.io/v1alpha1
kind: VolumeMigration
metadata:
  name: vol-mig
spec:
  migratedVolume:
  - sourcePvc: src-pvc1
    destinationPvc: dest-pvc1
    reclaimPolicySourcePvc: Delete
  - sourcePvc: scr-pvc2
    destinationPvc: dest-pvc2
    reclaimPolicySourcePvc: Delete
status:
  rejectedVolumeMigration:
    rejectedVolumes:
    - sourcePvc: src-pvc1
    - sourcePvc: src-pvc2
    reason: "Volumes needs to belong to the same VMI"
  phase: Rejected #(or Failed)
  phaseTransitionTimestamps:
  - phase: Rejected
    phaseTransitionTimestamp: "2024-01-12T08:54:11Z"
```

Reply:
Same please put the examples in the document, it is helpful later.

Reply:
Updated the API and added a similar example